File: src/main/java/com/weeth/domain/user/domain/entity/Cardinal.java
package com.weeth.domain.user.domain.entity;

import jakarta.persistence.*;
import com.weeth.domain.user.application.dto.request.CardinalUpdateRequest;
import com.weeth.domain.user.domain.entity.enums.CardinalStatus;
import com.weeth.global.common.entity.BaseEntity;
import lombok.AccessLevel;
import lombok.Builder;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.experimental.SuperBuilder;

@Entity
@Getter
@NoArgsConstructor(access = AccessLevel.PROTECTED)
@SuperBuilder
public class Cardinal extends BaseEntity {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(name = "cardinal_id")
    private Long id;

    @Column(unique = true, nullable = false)
    private Integer cardinalNumber;

    private Integer year;

    private Integer semester;

    @Builder.Default
    @Enumerated(EnumType.STRING)
    CardinalStatus status = CardinalStatus.DONE;

    public void update(CardinalUpdateRequest dto) {
        this.year = dto.year();
        this.semester = dto.semester();
    }

    public void inProgress() {
        this.status = CardinalStatus.IN_PROGRESS;
    }

    public void done() {
        this.status = CardinalStatus.DONE;
    }

}

File: src/main/kotlin/com/weeth/domain/user/domain/entity/Cardinal.kt
package com.weeth.domain.user.domain.entity

import com.weeth.domain.user.domain.entity.enums.CardinalStatus
import com.weeth.global.common.entity.BaseEntity
import jakarta.persistence.Column
import jakarta.persistence.Entity
import jakarta.persistence.EnumType
import jakarta.persistence.Enumerated
import jakarta.persistence.GeneratedValue
import jakarta.persistence.GenerationType
import jakarta.persistence.Id

@Entity
class Cardinal(
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(name = "cardinal_id")
    val id: Long = 0L,
    @Column(unique = true, nullable = false)
    val cardinalNumber: Int,
    var year: Int? = null,
    var semester: Int? = null,
    @Enumerated(EnumType.STRING)
    var status: CardinalStatus = CardinalStatus.DONE,
) : BaseEntity() {
    fun update(
        year: Int,
        semester: Int,
    ) {
        this.year = year
        this.semester = semester
    }

    fun inProgress() {
        status = CardinalStatus.IN_PROGRESS
    }

    fun done() {
        status = CardinalStatus.DONE
    }

    companion object {
        fun create(
            cardinalNumber: Int,
            year: Int? = null,
            semester: Int? = null,
            status: CardinalStatus = CardinalStatus.DONE,
        ): Cardinal =
            Cardinal(
                cardinalNumber = cardinalNumber,
                year = year,
                semester = semester,
                status = status,
            )
    }
}
